### PR TITLE
let users configure private IP of the internal LB

### LINKED ISSUE #5264: Make internal LB IP configurable
CAPZ assigns a default IP of `10.0.0.100` to the internal Load Balancer. We would like to make this configurable by the user. 
Also add validation to the webhooks validating _newly configured Internal LB IP_ belongs to the Control plane Subnet CIDR.
```[tasklist]
### Tasks
- [x] Create an optional job at K8s-test-infra to test APIServer ILB changes.
- [x] Create an e2e test at runs the new APIServerILB templates
- [x] Create APIServerILB tempalates for prow/ci
```

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- This PR is needed for updating Tilt workflow to spin up VM based templates with custom, non-overlapping Private IPs. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5264 

**Special notes for your reviewer**:

- This PR follows https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5311 .
- **Note:** This PR 
  -  updates defaulting webhook to create and assign a default internal IP when _APIServer ILB Feature flag enabled_
  - updates validation webhook to validate the internal IP when when _APIServer ILB Feature flag enabled_
  - updates defaulting webhook unit tests: preserves existing functionality/unit tests and adds new tests for feature flag scenarios.
  -  updates validating webhook unit tests: preserves existing functionality/unit tests and adds new tests for feature flag scenarios.
  - updates apiserver ilb template with custom CIDRs
  - updates apiserver ilb windows template with custom CIDRs
  - creates ci template for api server ILB for linux nodes
  - adds e2e tests for apiserver ilb feature flag 

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] cherry-pick candidate

**Release note**:

```release-note
feat APIServerILB: private IP of the internal LB can be customized
```

### CODE REVIEW

File: api/v1beta1/azurecluster_validation.go [hunk 1]
```diff
@@ -412,37 +412,56 @@ func validateAPIServerLB(lb *LoadBalancerSpec, old *LoadBalancerSpec, cidrs []st
...
+	// internal LB should not have a public IP.
+	if lb.Type == Internal {
+		if publicIPCount != 0 {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("frontendIPConfigs").Index(0).Child("publicIP"),
+				"Internal Load Balancers cannot have a Public IP"))
+		}
+		if privateIPCount != 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("frontendIPConfigs"), lb.FrontendIPs,
+				"API Server Load balancer of type private should have 1 frontend private IP"))
+		} else {
```
Comment (willie-yao) [inline api/v1beta1/azurecluster_validation.go:454]: Is there a possibility that privateIPCount = 0 here and does the following logic account for that?

File: api/v1beta1/azurecluster_validation_test.go [hunk 2]
```diff
@@ -1082,12 +1221,67 @@ func TestValidateAPIServerLB(t *testing.T) {
...
+				Field:    "apiServerLB.frontendIPConfigs[0].privateIP",
+				BadValue: "20.1.2.3",
+				Detail:   "Internal LB IP address needs to be in control plane subnet range ([10.0.0.0/24 10.1.0.0/24])",
+			},
+		},
 	}
 
 	for _, test := range testcases {
 		t.Run(test.name, func(t *testing.T) {
-			t.Parallel()
```
Comment (willie-yao) [inline api/v1beta1/azurecluster_validation_test.go:1089]: Was this meant to be removed?

File: templates/flavors/apiserver-ilb/patches/kubeadm-config-template.yaml [hunk 3]
```diff
@@ -8,6 +8,5 @@ spec:
       # /etc/hosts file is updated with a pre-created DNS name of the API server and internal load-balancer's IP.
       # This custom DNS Resolution of the API server ensures that the worker nodes can reach the API server when
       # the public IP of the API server is not accessible.
-      # 10.0.0.100 is the default IP that gets assigned to an internal load balancer.
       preKubeadmCommands:
-        - echo '10.0.0.100   ${CLUSTER_NAME}-${APISERVER_LB_DNS_SUFFIX}.${AZURE_LOCATION}.cloudapp.azure.com' >> /etc/hosts
+        - echo '30.0.0.100   ${CLUSTER_NAME}-${APISERVER_LB_DNS_SUFFIX}.${AZURE_LOCATION}.cloudapp.azure.com' >> /etc/hosts
```
Comment (willie-yao) [inline templates/flavors/apiserver-ilb/patches/kubeadm-config-template.yaml:12]: For my own understanding, why is this changed to 30?

File: test/e2e/azure_test.go [hunk 4]
```diff
@@ -1153,4 +1153,107 @@ var _ = Describe("Workload cluster creation", func() {
...
+	Context("Creating a self-managed VM based cluster using API Server ILB feature gate using default template [OPTIONAL][API-Server-ILB]", func() {
+		It("with three controlplane node and three worker nodes", func() {
+			clusterName = getClusterName(clusterNamePrefix, "apiserver-ilb")
+
+			// Set the environment variables required for the API Server ILB feature gate
+			Expect(os.Setenv("EXP_APISERVER_ILB", "true")).To(Succeed())
+			// Expect(os.Setenv("AZURE_INTERNAL_LB_PRIVATE_IP", "40.0.0.100")).To(Succeed())
+			// Expect(os.Setenv("AZURE_VNET_CIDR", "40.0.0.0/8")).To(Succeed())
+			// Expect(os.Setenv("AZURE_CP_SUBNET_CIDR", "40.0.0.0/16")).To(Succeed())
+			// Expect(os.Setenv("AZURE_NODE_SUBNET_CIDR", "40.1.0.0/16")).To(Succeed())
```
Comment (willie-yao) [inline test/e2e/azure_test.go:1166]: Should these be uncommented or removed?

File: test/e2e/azure_apiserver_ilb.go [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+                                             v
++-------------------------------------------+--------------------------------------------+
+| Validate Test Results                     | - Check All Pods Pass Tests                |
+| - Retry if Any Pod Fails                  | - Ensure Consistent Results                |
++-------------------------------------------+--------------------------------------------+
+                                             |
+                                             v
+                                  +--------------------------------+
+                                  |        Test Complete           |
+                                  +--------------------------------+
```
Comment (willie-yao) [inline test/e2e/azure_apiserver_ilb.go:82]: I love this comment describing what the test does! Kudos!

File: test/e2e/azure_apiserver_ilb.go [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+	groupName := os.Getenv(AzureResourceGroup)
+	fmt.Fprintf(GinkgoWriter, "Azure Resource Group: %s\n", groupName)
+	internalLoadbalancerName := fmt.Sprintf("%s-%s", input.ClusterName, "public-lb-internal")
+
+	backoff := wait.Backoff{
+		Duration: 300 * time.Second,
+		Factor:   0.5,
+		Jitter:   0.5,
+		Steps:    7,
+	}
```
Comment (willie-yao) [inline test/e2e/azure_apiserver_ilb.go:156]: nit: Might be good to set these to these established variables: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_privatecluster.go#L179

Comment (willie-yao) [reply test/e2e/azure_apiserver_ilb.go:156]: If you need a different value for them, it could be good to at least set it as a new variable

File: test/e2e/azure_apiserver_ilb.go [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+						},
+					},
+				},
+			},
+		},
+	}
+	err = workloadClusterClient.Create(ctx, nodeDebugDS)
+	Expect(err).NotTo(HaveOccurred())
+
+	backoff = wait.Backoff{
```
Comment (willie-yao) [inline test/e2e/azure_apiserver_ilb.go:306]: same as above!

File: test/e2e/azure_apiserver_ilb.go [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+		azureCluster.UnstructuredContent(),
+		deployedAzureCluster,
+	)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Getting the controlplane endpoint name")
+	controlPlaneEndpointDNSName, apiServerILBPrivateIP := "", ""
+	for _, frontendIP := range deployedAzureCluster.Spec.NetworkSpec.APIServerLB.FrontendIPs {
+		if frontendIP.PublicIP != nil && frontendIP.PublicIP.DNSName != "" {
+			fmt.Fprintf(GinkgoWriter, "Control Plane Endpoint Name: %s\n", frontendIP.PublicIP.DNSName)
```
Comment (willie-yao) [inline test/e2e/azure_apiserver_ilb.go:212]: nit: Might be good to use Logf instead of fmt.Fprintf here, as well as in the rest of the file.

File: test/e2e/azure_apiserver_ilb.go [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+			default:
+				return false, fmt.Errorf("node-debug pod %s is in an unexpected phase: %v", nodeDebugPod.Name, nodeDebugPod.Status.Phase)
+			}
+
+			listOfCommands := map[string][]string{
+				"Hello from node-debug pod": {"sh", "-c", "echo \"Hello from node-debug pod\""},
+				apiServerILBPrivateIP:       {"sh", "-c", "test -f /host/etc && cat /host/etc || echo 'File not found'"}, // /etc/host is mounted as /host/etc/hosts in the node-debug pod
+			}
+			testResult := map[string]bool{
+				"Hello from node-debug pod": false,
```
Comment (willie-yao) [inline test/e2e/azure_apiserver_ilb.go:385]: nit: Might be good to save "Hello from node-debug pod" to a variable since it's used in multiple places

File: api/v1beta1/azurecluster_default.go [hunk 10]
```diff
@@ -245,6 +247,29 @@ func (c *AzureCluster) setAPIServerLBDefaults() {
 				},
 			}
 		}
+		// if the API Server ILB feature is enabled, we should create a default internal LB IP
```
[the remark this answers is not in the document]
Comment (nawazkh) [reply api/v1beta1/azurecluster_default.go:250]: Ack. 
Updating it to 
```
// If the API Server ILB feature is enabled, create a default internal LB IP or use the specified one
```

File: api/v1beta1/azurecluster_validation.go [hunk 11]
```diff
@@ -412,37 +412,56 @@ func validateAPIServerLB(lb *LoadBalancerSpec, old *LoadBalancerSpec, cidrs []st
...
+		if lb.FrontendIPs[i].PrivateIPAddress != "" {
+			privateIPCount++
+			privateIP = lb.FrontendIPs[i].PrivateIPAddress
+		}
+	}
+	if lb.Type == Public {
+		// there should be one public IP for public LB.
+		if publicIPCount != 1 || ptr.Deref[int32](lb.FrontendIPsCount, 1) != 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("frontendIPConfigs"), lb.FrontendIPs,
+				"API Server Load balancer should have 1 Frontend IP"))
```
[the remark this answers is not in the document]
Comment (nawazkh) [reply api/v1beta1/azurecluster_validation.go:430]: FrontendIPsCount translates to creation of publicIPs in here https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/61c6f3f180539609090fbf5a943b2e021cfbf129/api/v1beta1/azurecluster_default.go#L371-L382 
Checking both the counts implies the same and hence the reduced error message. 

Shall we make the error message more verbose ?

File: api/v1beta1/azurecluster_validation.go (hunk 1 above)
Comment (nawazkh) [reply api/v1beta1/azurecluster_validation.go:454]: It is possible for a private LB of type Internal to have `privateIPCount = 0`. 
And yes, if `privateIPCount == 0` then the highlighted code block is the logic flow we want the control to execute
```
if privateIPCount != 1 {
	allErrs = append(allErrs, field.Invalid(fldPath.Child("frontendIPConfigs"), lb.FrontendIPs,
		"API Server Load balancer of type private should have 1 frontend private IP"))
}
```

File: api/v1beta1/azurecluster_validation_test.go (hunk 2 above)
Comment (nawazkh) [reply api/v1beta1/azurecluster_validation_test.go:1089]: Yep, `t.Parallel()` was meant to be removed when working with global feature flag. Ref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e0f2b44aec58ed04bbd754a98bace470bea32fff/api/v1beta1/azurecluster_validation_test.go#L1282-L1284

Having `t.Parallel()` when setting global values will mess with the test leading to erroneous output or race conditions( race condition not applicable in our scenario). 
Hence, I had to remove `t.Parallel()`.

File: templates/flavors/apiserver-ilb/patches/kubeadm-config-template.yaml (hunk 3 above)
Comment (nawazkh) [reply templates/flavors/apiserver-ilb/patches/kubeadm-config-template.yaml:12]: `10.0.0.100` is the default value for internal ILB IP Address. Ref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e0f2b44aec58ed04bbd754a98bace470bea32fff/api/v1beta1/azurecluster_default.go#L45

A non default value, `30.0.0.100`, will help in spotting runtime or user errors. 
Also, using `30.0.0.100` supports the goal of this PR; lets user configure the IP of the ILB.

File: test/e2e/azure_test.go (hunk 4 above)
Comment (nawazkh) [reply test/e2e/azure_test.go:1166]: Good spot, they should be removed

File: test/e2e/azure_apiserver_ilb.go (hunk 6 above)
Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go:156]: Thanks for the linking https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_privatecluster.go#L179. 

I purposefully created a new local variable, `backoff`,  with my own _**readable values**_ because the hardcoded values, mentioned below, https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/61c6f3f180539609090fbf5a943b2e021cfbf129/test/e2e/azure_privatecluster.go#L178-L183 
are not readable. 

For instance, unless I click and navigate to `retryBackoffInitialDuration` declaration, I do not know that it is set to `100 * time.Millisecond`. 

The better approach would be to rename `retryBackoffInitialDuration` -> `retryBackoff_100MilSeconds` or something that makes reading the code easier.

Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go:156]: > If you need a different value for them, it could be good to at least set it as a new variable

Can you please elaborate ?
Would you like me to create a new local variable for the Duration values ? 
Or rename the variable `backoff` ? 
Since `backoff` is a locally scoped variable that is usually changing per retry func, do we really need to standardize the values for Duration, Jitter... etc ?

File: test/e2e/azure_apiserver_ilb.go (hunk 7 above)
Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go:306]: Same doubt as above https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/5332#discussion_r1882952753

File: test/e2e/azure_apiserver_ilb.go (hunk 9 above)
Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go:385]: Good idea, will do.

File: test/e2e/azure_apiserver_ilb.go (hunk 8 above)
Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go:212]: `fmt.Fprintf(GinkgoWriter, .....)` helped me log the output inline with the test. 
Do you know if `Logf()` achieves the same ?

File: api/v1beta1/azurecluster_default.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (nawazkh) [reply api/v1beta1/azurecluster_default.go]: Added via f875002db330906ad0735b66246ef348a78f0857

File: api/v1beta1/azurecluster_default.go [hunk 12]
```diff
@@ -245,6 +247,29 @@ func (c *AzureCluster) setAPIServerLBDefaults() {
...
+						lb.FrontendIPs[i].Name = generateFrontendIPConfigName(lb.Name) + "-internal-ip"
+					}
+					privateIPFound = true
+					break
+				}
+			}
+			// if no private IP is found, we should create a default internal LB IP
+			if !privateIPFound {
+				privateIP := FrontendIP{
+					Name: generateFrontendIPConfigName(lb.Name) + "-internal-ip",
```
Comment (Jont828) [inline api/v1beta1/azurecluster_default.go:265]: Maybe add a function parameter like `internal` or make a new function here, since we have to generate the name like this here and in 256.

File: test/e2e/azure_test.go (hunk 4 above)
Comment (nawazkh) [reply test/e2e/azure_test.go]: Removed via f875002db330906ad0735b66246ef348a78f0857

File: test/e2e/azure_apiserver_ilb.go (hunk 9 above)
Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go]: Added via f875002db330906ad0735b66246ef348a78f0857

File: api/v1beta1/azurecluster_validation.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jont828) [reply api/v1beta1/azurecluster_validation.go:430]: Maybe change it to `public API server load balancer must exactly 1 frontend IP`?

File: api/v1beta1/azurecluster_validation.go [hunk 13]
```diff
@@ -412,37 +412,56 @@ func validateAPIServerLB(lb *LoadBalancerSpec, old *LoadBalancerSpec, cidrs []st
...
 
+	// internal LB should not have a public IP.
+	if lb.Type == Internal {
+		if publicIPCount != 0 {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("frontendIPConfigs").Index(0).Child("publicIP"),
+				"Internal Load Balancers cannot have a Public IP"))
+		}
+		if privateIPCount != 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("frontendIPConfigs"), lb.FrontendIPs,
+				"API Server Load balancer of type private should have 1 frontend private IP"))
```
Comment (Jont828) [inline api/v1beta1/azurecluster_validation.go:453]:
```suggestion
				"API Server Load balancer of type private must have 1 frontend private IP"))
```

To match the language of `cannot` in the other errors.

**Review by Jont828 — COMMENTED** [review]
LGTM pending nits and Willie's comments

File: api/v1beta1/azurecluster_validation.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (willie-yao) [reply api/v1beta1/azurecluster_validation.go:430]: If it implies the same, then the error message is sufficient. I think it might help to add a comment saying that the two checks are for the same purpose though. I was just a bit confused since two different values were checked, but the error only referred to one of them.

File: test/e2e/azure_apiserver_ilb.go (hunk 6 above)
Comment (willie-yao) [reply test/e2e/azure_apiserver_ilb.go:156]: Okay that's fair! There isn't a need to standardize them, and if it is more readable the way you have it, I think it's best to leave it as-is.

File: test/e2e/azure_apiserver_ilb.go (hunk 8 above)
Comment (willie-yao) [reply test/e2e/azure_apiserver_ilb.go:212]: It should achieve the same and I've always used Logf to print stuff in-line. Logf under the hood is fmt.Fprintf as seen here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/61c6f3f180539609090fbf5a943b2e021cfbf129/test/e2e/log.go#L36

It seems like Logf is the preferred way to print stuff in E2E tests as it prints the log level and timestamp

File: api/v1beta1/azurecluster_default.go (hunk 12 above)
Comment (nawazkh) [reply api/v1beta1/azurecluster_default.go:265]: adding it

File: api/v1beta1/azurecluster_validation.go (hunk 13 above)
Comment (nawazkh) [reply api/v1beta1/azurecluster_validation.go:453]: I copied over the error message from the earlier implementation, so I would probably leave it as is to preserve any back comp issues. 
Although I think changing `should` -> `must` wont lead to any comp issues.

File: api/v1beta1/azurecluster_validation.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (nawazkh) [reply api/v1beta1/azurecluster_validation.go:431]: Added via : 57e92745093fb5ca4a408e5dc097e7195eae557f

File: api/v1beta1/azurecluster_default.go (hunk 12 above)
Comment (nawazkh) [reply api/v1beta1/azurecluster_default.go]: Added via 57e92745093fb5ca4a408e5dc097e7195eae557f

**Comment by nawazkh** [comment]
```yaml
RESPONSE 400: 400 Bad Request
	ERROR CODE: PublicIPCountLimitReached
	--------------------------------------------------------------------------------
	{
	  "error": {
	    "code": "PublicIPCountLimitReached",
	    "message": "Cannot create more than 100 public IP addresses for this subscription in this region.",
	    "details": []
	  }
	}

```

File: test/e2e/azure_apiserver_ilb.go (hunk 8 above)
Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go:212]: Updating

Comment (nawazkh) [reply test/e2e/azure_apiserver_ilb.go]: Updated, please take a look.

**Review by willie-yao — APPROVED** [review]
/lgtm

Awesome work @nawazkh! This is a huge fix for CAPZ 🚀

**Comment by Jont828** [comment]
/approve

Nice work!

File: api/v1beta1/azurecluster_default_test.go [hunk 14]
```diff
@@ -1375,12 +1475,135 @@ func TestAPIServerLBDefaults(t *testing.T) {
...
+					},
+				},
+			},
+		},
 	}
 
 	for _, c := range cases {
 		tc := c
 		t.Run(tc.name, func(t *testing.T) {
-			t.Parallel()
```
Comment (mboersma) [inline api/v1beta1/azurecluster_default_test.go:1383]: This didn't work with `t.Parallel`?

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

Ignore my comment about `t.Parallel`; I saw your answer to @willie-yao about the same thing. 👍

File: templates/flavors/apiserver-ilb/patches/kubeadm-config-template.yaml (hunk 3 above)
Comment (jackfrancis) [reply templates/flavors/apiserver-ilb/patches/kubeadm-config-template.yaml:12]: if we wanna demonstrate a non-default private IP address value, we want to choose a private IP address

Ref:

- https://www.iana.org/help/private-addresses

(unless I'm misunderstanding something)
